En passant

steve-mcclellan opened the pull request:
This branch adds the ability for a pawn to capture another pawn [en passant](https://en.wikipedia.org/wiki/En_passant). (In short, a pawn that advances two squares from its initial position may, on the very next move, be captured by an enemy pawn as though the former had advanced only one square.)

![anim](https://cloud.githubusercontent.com/assets/17390227/15801863/57d0c940-2a70-11e6-88c9-f74173b519c8.gif)

This branch:
- Adds an en_passant_file column to the games table, which tracks whether an en passant capture is legal, and updates this value in the move_to! method.
- Updates the Pawn class's valid_move? helper methods.
- Creates a Pawn-class move_to! method, which executes the Piece model's move_to! method, and removes an en-passant-captured pawn, if applicable.

steve-mcclellan committed:
Add en_passant_file column to the games table

Here, "file" is used as the chess-ier synonym for "column".

An en passant capture is only possible on the move immediately following
a two-square advance by an opposing pawn. This column keeps track of
whether or not this applies.

At each move, the appropriate move_to! method will update this value to
the correct file/column number if the move is a two-square advance by
a pawn, and set it to nil/null otherwise.

If an en passant capture is attempted, valid_move? and move_to! (or
their helper methods) will read this value as part of the processes of
verifying the move's legality and capturing the enemy pawn.

steve-mcclellan committed:
Add nth_rank helper method to piece model

As noted in the method comment, this method takes the rank from the
current player's perspective, and returns the corresponding y value.
For example, pawns start on the player's second rank, which is 2 for
White and 7 for Black. Therefore, nth_rank(2) returns 2 if the piece is
white, and 7 if it's black.

steve-mcclellan committed:
Refactor valid_double_move? to use nth_rank

The valid_double_move? method, in the Pawn model, verifies that the pawn
is on its original row, and that neither of the spaces immediately in
front of it is occupied.

The newly-created nth_rank helper method removes the need to explicitly
call the ternary operator to set the values of origin_row and
passed_row.

Because each of these values is used only once, and they can
now be defined non-verbosely, the variables themselves are no longer
particularly useful, and the calls to nth_rank can be put directly in
the methods which use the values.

(Additionally, the method's last two lines are refactored into one.)

steve-mcclellan committed:
Update valid_capture? method for en passant

Previously, the Pawn model's valid_capture method simply returned true
if the destination square contained an enemy piece, and false otherwise.

With this update, the method continues to return true if a standard
capture is possible, but also returns true if an enemy pawn can be
captured en passant. This will be the case if:

- the capturing pawn starts on its fifth rank, and
- the game's immediately-preceding move was a pawn advancing two squares
  along the destination file (column).

If the latter applies, the game's newly-created en_passant_file
attribute will equal the x coordinate of the destination square.

steve-mcclellan committed:
Add pawn-specific move_to! method

Create a move_to! method in the Pawn model which overrides (and then
calls, via "super") the Piece model's move_to! method.

The Pawn-specific variety destroys a pawn that has been captured en
passant (via the capture_en_passant! helper method), and updates the
game's en_passant_file attribute with the file number (x_coord) if the
current move is a two-square advance (thus leaving the currently-moving
pawn vulnerable to en passant capture on the next move), and nil
otherwise.

After the call to super, which handles standard captures and updates the
coordinates of the moving piece, a commented-out call to the
not-yet-created promote! method is added as a placeholder.

The capture_en_passant! helper method checks whether:

- the current move is diagonal (indicating a capture),
- the move is to the piece's sixth rank (the enemy's third rank, which
  is always the row passed over by a two-square pawn advance), and
- the game's last move was a two-square pawn advance on the same column
  as the current move's destination (which would have caused the game's
  en_passant_file attribute to be set to that column number).

If all of those apply, an en_passant capture is occurring. The captured
piece will have the same x_coord as the current move's destination, and
the same y_coord that the capturing pawn had *before* the move. The
piece on that square is then destroyed.

steve-mcclellan committed:
Update en_passant_file in Piece's move_to! method

If the current move is not being made by a pawn, en passant capture is
not possible on the next move. Set the game's en_passant_file attribute
to nil to reflect this. (For moves by pawns, the move_to! method in the
Pawn class sets this value appropriately.)

This commit also includes the following minor tweaks:
- Reload the game before retrieving en_passant_file in the pawn's
  valid_capture? method.
- Remove a commented-out line of debugging code.

steve-mcclellan committed:
Fix typo in comment

h35514n commented on db/migrate/20160603182320_add_en_passant_file_to_games.rb:
```diff
@@ -0,0 +1,5 @@
+class AddEnPassantFileToGames < ActiveRecord::Migration
+  def change
+    add_column :games, :en_passant_file, :integer
```
I'm a bit confused by the naming here -- does `file` refer to a position on the board, as in "rank and file"? (My knowledge of chess might be sketchier than I thought 😄)

h35514n replied:
Ah, I see that [it is](https://github.com/letmecheck/app/pull/56/commits/d4fd906f54db0089d127b08b29fb66777660a6e1#diff-1b541d09ba44dabaa60af6e62f436119R26). I think I would have expected a boolean (e.g., `en_passant_eligible`) here rather than an integer. Is the current `en_passant_file` (the integer, that is) used for anything during gameplay?

h35514n replied:
Lol, I see that [it is](https://github.com/letmecheck/app/pull/56/commits/d4fd906f54db0089d127b08b29fb66777660a6e1#diff-1b541d09ba44dabaa60af6e62f436119R61) 😄 It might be a good idea to record the strategy for computing whether or not it's a valid move in the commit message here, and adding something like an `#en_passant_eligible?` method wrapping the validation logic involved.

steve-mcclellan committed:
Add basic test coverage

Set up a game, make a series of moves leading to an en passant capture,
and verify that the capture is valid and the captured piece is
destroyed.

Then, make a series of moves that would lead to the possibility of en
passant capture, but make another move next instead. Verify that the
en passant capture is invalid, and that the no-longer-vulnerable pawn
is still present.

steve-mcclellan committed:
Add non-pawn and no-two-square-advance tests

Verify that en passant capture cannot be made by a non-pawn, by
attempting to take a pawn that just advanced two squares by moving
a bishop to the square it passed through. This is a legal bishop move,
but leaves the pawn uncaptured.

Verify that en passant capture cannot be made on a pawn that has made
two separate one-square advances, rather than a single two-square
advance.

steve-mcclellan committed:
Add tests for bug that appeared during rebase

While this branch was being prepared, the Piece model's move_to! method
was changed to reject the move unless valid_move? returns true.

This created undesired behavior, as en passant's current implementation
calls the Piece model's move_to! method (to move the capturing pawn)
after resetting the game's en_passant_file attribute. This called
valid_move? on the capture, which returned false because of this updated
value.

The net result of this was that when a pawn made an en passant capture,
the captured pawn was destroyed, but the move was then flagged as
invalid, and the capturing pawn snapped back to its original position.

This commit expands the spec to verify that the capturing pawn has the
proper coordinates after the capture. It also verifies that successful
en passant capture works as expected on pawns of each color.

steve-mcclellan committed:
Ensure that move isn't rejected after e.p. capture

Rearrange helper method calls in Pawn's move_to! method, so that no
action occurs until after the call to super (and its call to
valid_move?), and execution is halted if super rejects the move.

Also fix a typo in the spec.

bennypaulino commented:
:shipit: :+1:
